lib/engines/pki/pki.ex

```diff
@@ -0,0 +1,221 @@
+defmodule Ptolemy.Engines.PKI do
+    @moduledoc """
+    `Ptolemy.Engine.PKI` provides a public facing API for CRUD operations using the config file
+  
+    All functions start with an additional c implies the function uses the secret configuration
+  
+    The other functions act as a mask to the appropriate functions in the engine
+    """
+  
+      
+    alias Ptolemy.Engines.PKI.Engine
+    alias Ptolemy.Server
+    @doc """
+    Create a role from the specification provided
+      
+    ## Example
+    ```elixir
+    iex(2)> Ptolemy.Engines.PKI.create(server, :pki_engine1, :test_role1, %{allow_any_name: true})
+    ```
+    """
+    @spec create(pid(), atom(), atom(), map()) :: {:ok | :error, String.t()}
+    def create(pid, engine_name, role, payload \\ %{}) do
+      path = get_pki_path!(pid, engine_name, role, "roles")
+      path_create(pid, path, payload)
+    end
+
+    @spec create!(pid(), atom(), atom(), map()) :: :ok
+    def create!(pid, engine_name, role, payload \\ %{}) do
+      case create(pid, engine_name, role, payload) do
+        {:error, msg} -> raise RuntimeError, message: msg
+        _resp -> :ok
+      end
+    end
+    
+    @spec path_create(pid(), String.t(), map()) :: {:ok | :error, String.t()}
+    def path_create(pid, path, payload \\ %{}) do
+      client = create_client(pid)
+      Engine.create_role(client, path, payload)
+    end
+    
+    @doc """
+    Generate a certificate from the role
+      
+    ## Example
+    ```elixir
+    iex(2)> Ptolemy.Engines.PKI.read(server, :pki_engine1, :test_role1, "www.example.com")
+    ```
+    """
+    @spec read(pid(), atom(), atom(), String.t(), map()) :: {:ok | :error, String.t()}
+    def read(pid, engine_name, role, common_name, payload \\ %{}) do
+      path = get_pki_path!(pid, engine_name, role, "issue")
+      path_read(pid, path, common_name, payload)
+    end
+    
+    @doc """
+    The same as read(), except it raises an exception when error occurs like all bang functions
+    """
+    @spec read!(pid(), atom(), atom(), String.t(), map()) :: :ok
+    def read!(pid, engine_name, role, common_name, payload \\ %{}) do
+      case read(pid, engine_name, role, common_name, payload) do
+        {:error, msg} -> raise RuntimeError, message: msg
+        resp -> resp
+      end
+    end
+
+    @spec read(pid(), String.t(), String.t(), map()) :: {:ok | :error, String.t()}
+    def path_read(pid, path, common_name, payload) do
```

Comment:
the @spec function name above this definition is different than the actual function name. probably just a typo